envoy/upstream/outlier_detection.h

```diff
@@ -47,6 +47,26 @@ enum class Result {
   ExtOriginRequestSuccess // Request was completed successfully.
 };
 
+// Types of outlier detection extension results which can be reported.
+enum class ExtResultType;
+
+/*
+ * Class carries result of a transaction with upstream entity
+ * or generated internally by Envoy.
+ * Different categories of results will be derived from that base class.
+ * Those categories of results are fed only into Outlier Detection extensions.
+ */
+class ExtResult {
+public:
+  ExtResult() = delete;
+  ExtResult(ExtResultType type) : type_(type) {}
+  virtual ExtResultType type() const { return type_; };
+  virtual ~ExtResult() = default;
+
+private:
+  const ExtResultType type_;
+};
```

Comment:
That is the base class for different types of results. So, Http results will be reported via a class derived from ExtResult. This is needed because when matching to user-defined buckets happens, the type is checked first. So, Http codes are matched only versus a bucket which has the same type. It does not make sense to match Http code against a bucket which has say database errors: https://github.com/envoyproxy/envoy/blob/43361bb3a0a9cf14a027abc822230e334f9f7072/source/extensions/outlier_detection_monitors/common/monitor_base_impl.cc#L34-L38

The only common property across all different results is type.